libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,567 @@
+
+				for (auto entry: block->entries)
+					toVisit.emplace_back(entry);
+			}
+			else
+				continue;
+		}
+
+		for (auto [block, target]: backwardsJumps)
+			if (ranges::any_of(
+				m_layout.blockInfos[target].entryLayout,
+				[exitLayout = m_layout.blockInfos[block].exitLayout](StackSlot const& _slot) {
+					return !util::findOffset(exitLayout, _slot);
+				}
+			))
+			{
+				// This block jumps backwards, but does not provide all slots required by the jump target on exit.
+				// Therefore we need to visit the subgraph between ``target`` and ``block`` again.
+				// In particular we can visit backwards starting from ``block`` and mark all entries to-be-visited-
+				// again until we hit ``target``.
+				toVisit.emplace_front(block);
+				util::BreadthFirstSearch<CFG::BasicBlock const*>{{block}}.run(
+					[&visited, target = target](CFG::BasicBlock const* _block, auto _addChild) {
+						visited.erase(_block);
+						if (_block == target)
+							return;
+						for (auto const* entry: _block->entries)
+							_addChild(entry);
+					}
+				);
+				// TODO: while the above is enough, the layout of ``target`` might change in the process.
+				// While the shuffled layout for ``target`` will be compatible, it can be worthwhile propagating
+				// it further up once more.
+				// This would mean not stopping at _block == target above or even doing visited.clear() here, revisiting the entire graph.
+				// This is a tradeoff between the runtime of this process and the optimality of the result.
+				// Also note that while visiting the entire graph again *can* be helpful, it can also be detrimental.
+				// Also note that for some reason using visited.clear() is incompatible with skipping the revisit
+				// of already seen exit layouts above, I'm not sure yet why.
+			}
+	}
+
+	stitchConditionalJumps(_entry);
+	fixStackTooDeep(_entry);
+}
+
+Stack StackLayoutGenerator::combineStack(Stack const& _stack1, Stack const& _stack2)
+{
+	if (_stack1.empty())
+		return _stack2;
+	if (_stack2.empty())
+		return _stack1;
+
+	// TODO: there is probably a better way than brute-forcing. This has n! complexity or worse, so
+	// we can't keep it like this.
+
+	Stack commonPrefix;
+	for (auto&& [slot1, slot2]: ranges::zip_view(_stack1, _stack2))
+	{
+		if (!(slot1 == slot2))
+			break;
+		commonPrefix.emplace_back(slot1);
+	}
+	Stack stack1Tail = _stack1 | ranges::views::drop(commonPrefix.size()) | ranges::to<Stack>;
+	Stack stack2Tail = _stack2 | ranges::views::drop(commonPrefix.size()) | ranges::to<Stack>;
+
+	Stack candidate;
+	for (auto slot: stack1Tail)
+		if (!util::findOffset(candidate, slot))
+			candidate.emplace_back(slot);
+	for (auto slot: stack2Tail)
+		if (!util::findOffset(candidate, slot))
+			candidate.emplace_back(slot);
+	cxx20::erase_if(candidate, [](StackSlot const& slot) {
+		return holds_alternative<LiteralSlot>(slot) || holds_alternative<FunctionCallReturnLabelSlot>(slot);
+	});
+
+	std::map<size_t, Stack> sortedCandidates;
```

Comment:
Interestingly, the main thing the rest of this does is to balance out the stack shuffling between conditional jump targets.
Just returning ``candidate`` right here can be better and filling in ``stack2Tail`` before ``stack1Tail`` above and then just returning ``candidate`` here can be better in some cases.
I'd actually consider to introduce ``likely`` and ``unlikely`` branches like e.g. the linux kernel does...
For example branches leading to ``panic`` I would classify "unlikely", branches like the huge ``if iszero(lt(calldatasize(), 4))`` branch in the dispatch would be "likely". Then I could try to balance by default or favour a particular branch, if the if is marked likely or unlikely.

Reply:
Generally, if anyone has a non-exponential way to do this, that would of course also be great :-).